File: epi_judge_java/epi/NumberOfScoreCombinations.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import java.util.List;
public class NumberOfScoreCombinations {
  @EpiTest(testDataFile = "number_of_score_combinations.tsv")

  public static int
  numCombinationsForFinalScore(int finalScore,
                               List<Integer> individualPlayScores) {
    return epi.kashyap.NumberOfScoreCombinationsKt.numCombinationsForFinalScore(finalScore, individualPlayScores);
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "NumberOfScoreCombinations.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_java/epi/kashyap/NumberOfScoreCombinations.kt
package epi.kashyap

fun numCombinationsForFinalScore(finalScore: Int, individualPlayScores: List<Int>): Int {
    val combinations = Array<Int>(finalScore + 1, {0})
    combinations[0] = 1

    individualPlayScores.forEach { score ->
        for (i in combinations.indices) {
            if (combinations[i] == 0) continue
            val next = i + score
            if (next in combinations.indices) {
                combinations[next] += combinations[i]
            }
        }
    }
    return combinations[finalScore]
}

fun main(args: Array<String>) {
    println(numCombinationsForFinalScore(4, listOf(1, 2, 3)))
}